official/resnet/ctl/ctl_imagenet_main.py

```diff
@@ -0,0 +1,295 @@
+    train_steps = min(flags_obj.train_steps, steps_per_epoch)
+    train_epochs = 1
+
+  strategy_scope = distribution_utils.get_strategy_scope(strategy)
+  with strategy_scope:
+    logging.info('Building Keras ResNet-50 model')
+    model = resnet_model.resnet50(num_classes=imagenet_main.NUM_CLASSES,
+                                  dtype=dtype, batch_size=flags_obj.batch_size)
+
+    optimizer = tf.keras.optimizers.SGD(
+        learning_rate=_BASE_LEARNING_RATE, momentum=0.9, nesterov=True)
+
+    training_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
+        'training_accuracy', dtype=tf.float32)
+    test_loss = tf.keras.metrics.Mean('test_loss', dtype=tf.float32)
+    test_accuracy = tf.keras.metrics.SparseCategoricalAccuracy(
+        'test_accuracy', dtype=tf.float32)
+    logging.info('Finished building Keras ResNet-50 model')
+
+    def train_step(train_ds_inputs):
+      """Training StepFn."""
+      def step_fn(inputs):
+        """Per-Replica StepFn."""
+        images, labels = inputs
+        with tf.GradientTape() as tape:
+          logits = model(images, training=True)
+
+          prediction_loss = tf.keras.losses.sparse_categorical_crossentropy(
+              labels, logits)
+          loss1 = tf.reduce_sum(prediction_loss) * (1.0/ flags_obj.batch_size)
+          loss2 = tf.reduce_sum(model.losses) / strategy.num_replicas_in_sync
+          loss = loss1 + loss2
+
+        grads = tape.gradient(loss, model.trainable_variables)
+        optimizer.apply_gradients(zip(grads, model.trainable_variables))
+
+        training_accuracy.update_state(labels, logits)
+        return loss
+
+      if strategy:
+        per_replica_losses = strategy.experimental_run_v2(
+            step_fn, args=(train_ds_inputs,))
+        return strategy.reduce(tf.distribute.ReduceOp.SUM, per_replica_losses,
+                               axis=None)
+      else:
+        return step_fn(train_ds_inputs)
+
+    def test_step(test_ds_inputs):
+      """Evaluation StepFn."""
+      def step_fn(inputs):
+        images, labels = inputs
+        logits = model(images, training=False)
+        loss = tf.keras.losses.sparse_categorical_crossentropy(labels,
+                                                               logits)
+        loss = tf.reduce_sum(loss) * (1.0/ flags_obj.batch_size)
+        test_loss.update_state(loss)
+        test_accuracy.update_state(labels, logits)
+
+      if strategy:
+        strategy.experimental_run_v2(step_fn, args=(test_ds_inputs,))
+      else:
+        step_fn(test_ds_inputs)
+
+    if flags_obj.enable_function:
+      train_step = tf.function(train_step)
+      test_step = tf.function(test_step)
+
+    epoch_exp_per_sec = []
+    stats = {}
+    for epoch in range(train_epochs):
+      logging.info('Starting to run epoch: %s', epoch)
+      train_iterator = iter(train_ds)
+
+      step = 0
+      total_loss = 0.0
+      batch_exp_per_sec = []
+      for step in range(train_steps):
```

Comment:
any reason to not write this as the nicer version `for xyz in train_ds` ?

Reply:
There were outstanding issues when I started working on this PR so I chose this path. I intend to add this use case as well.

Reply:
i see. i thought most of the issues were with putting the iteration inside the tf.function, didn't know there were issues with it being outside too. 
Lets have multiple variations for the performance benchmarks. For the accuracy benchmark, we can just pick one variation perhaps? ideally it is one which handles last partial batch correctly so we get the eval on the entire dataset.

Reply:
I just tried making the change you suggested and ran into issue. Will follow up and have that version as part of these benchmarks.